File: src/com/mistraltech/smogen/codegenerator/matchergenerator/PsiTypeConverter.java
package com.mistraltech.smogen.codegenerator.matchergenerator;

import com.intellij.psi.PsiArrayType;
import com.intellij.psi.PsiClassType;
import com.intellij.psi.PsiPrimitiveType;
import com.intellij.psi.PsiType;
import com.intellij.psi.PsiTypeVisitor;
import com.intellij.psi.PsiWildcardType;
import com.intellij.psi.impl.source.PsiClassReferenceType;
import com.mistraltech.smogen.codegenerator.javabuilder.TypeBuilder;
import com.mistraltech.smogen.codegenerator.javabuilder.TypeParameterBuilder;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Map;

public class PsiTypeConverter extends PsiTypeVisitor<TypeBuilder> {
    private final Map<String, String> typeParameterMap;
    private final TypeBuilder typeBuilder = TypeBuilder.aType();
    private boolean boxed;

    public PsiTypeConverter(@NotNull Map<String, String> typeParameterMap) {
        this(true, typeParameterMap);
    }

    public PsiTypeConverter(boolean boxed, @NotNull Map<String, String> typeParameterMap) {
        this.boxed = boxed;
        this.typeParameterMap = typeParameterMap;
    }

    @Nullable
    @Override
    public TypeBuilder visitPrimitiveType(PsiPrimitiveType primitiveType) {
        typeBuilder.withName(boxed ? primitiveType.getBoxedTypeName() : primitiveType.getCanonicalText());
        return super.visitPrimitiveType(primitiveType);
    }

    @Nullable
    @Override
    public TypeBuilder visitArrayType(PsiArrayType arrayType) {
        boxed = false; // don't need to (or want to) box array types
        arrayType.getComponentType().accept(this);
        typeBuilder.withArrayDimensions(arrayType.getArrayDimensions());
        return super.visitArrayType(arrayType);
    }

    @Nullable
    @Override
    public TypeBuilder visitClassType(PsiClassType classType) {
        String name = (classType instanceof PsiClassReferenceType) ?
                ((PsiClassReferenceType) classType).getReference().getQualifiedName() :
                classType.getClassName();

        String mappedName = typeParameterMap.containsKey(name) ? typeParameterMap.get(name) : name;

        typeBuilder.withName(mappedName);

        for (PsiType param : classType.getParameters()) {
            PsiTypeParameterConverter converter = new PsiTypeParameterConverter(typeParameterMap);
            param.accept(converter);
            typeBuilder.withTypeBinding(converter.getTypeParameterBuilder());
        }

        return super.visitClassType(classType);
    }

    public TypeBuilder visitType(PsiType type) {
        return typeBuilder;
    }

    private static class PsiTypeParameterConverter extends PsiTypeVisitor<Object> {
        private final Map<String, String> typeParameterMap;
        private TypeParameterBuilder typeParameterBuilder = TypeParameterBuilder.aTypeParameter();

        PsiTypeParameterConverter(Map<String, String> typeParameterMap) {
            this.typeParameterMap = typeParameterMap;
        }

        @Nullable
        @Override
        public Object visitClassType(PsiClassType classType) {
            PsiTypeConverter typeConverter = new PsiTypeConverter(typeParameterMap);
            TypeBuilder typeBuilder = classType.accept(typeConverter);
            typeParameterBuilder.withType(typeBuilder);
            return super.visitClassType(classType);
        }

        @Nullable
        @Override
        public Object visitWildcardType(PsiWildcardType wildcardType) {
            if (wildcardType.getBound() != null) {
                wildcardType.getBound().accept(this);
            }

            typeParameterBuilder.withSubTypes(wildcardType.isExtends());
            typeParameterBuilder.withSuperTypes(wildcardType.isSuper());

            return super.visitWildcardType(wildcardType);
        }

        TypeParameterBuilder getTypeParameterBuilder() {
            return typeParameterBuilder;
        }
    }
}

File: src/main/kotlin/com/mistraltech/smogen/codegenerator/utils/PsiTypeConverter.kt
package com.mistraltech.smogen.codegenerator.utils

import com.intellij.psi.PsiArrayType
import com.intellij.psi.PsiClassType
import com.intellij.psi.PsiPrimitiveType
import com.intellij.psi.PsiType
import com.intellij.psi.PsiTypeVisitor
import com.intellij.psi.PsiWildcardType
import com.intellij.psi.impl.source.PsiClassReferenceType
import com.mistraltech.smogen.codegenerator.javabuilder.TypeBuilder
import com.mistraltech.smogen.codegenerator.javabuilder.TypeParameterBuilder

class PsiTypeConverter(private var boxed: Boolean, private val typeParameterMap: Map<String, String>) :
    PsiTypeVisitor<TypeBuilder>() {

    private val typeBuilder = TypeBuilder.aType()

    constructor(typeParameterMap: Map<String, String>) : this(true, typeParameterMap)

    override fun visitPrimitiveType(primitiveType: PsiPrimitiveType): TypeBuilder {
        val typeName = if (boxed) primitiveType.boxedTypeName else primitiveType.canonicalText
        typeBuilder.withName(typeName)

        return super.visitPrimitiveType(primitiveType)
    }

    override fun visitArrayType(arrayType: PsiArrayType): TypeBuilder {
        boxed = false // don't need to (or want to) box array types

        arrayType.componentType.accept(this)
        typeBuilder.withArrayDimensions(arrayType.arrayDimensions)

        return super.visitArrayType(arrayType)
    }

    override fun visitClassType(classType: PsiClassType): TypeBuilder {
        val name = if (classType is PsiClassReferenceType) classType.reference.qualifiedName else classType.className
        val mappedName = typeParameterMap.getOrDefault(name, name)
        typeBuilder.withName(mappedName)

        classType.parameters.forEach {
            val converter = PsiTypeParameterConverter(typeParameterMap)
            it.accept(converter)
            typeBuilder.withTypeBinding(converter.typeParameterBuilder)
        }

        return super.visitClassType(classType)
    }

    override fun visitType(type: PsiType): TypeBuilder {
        return typeBuilder
    }

    private class PsiTypeParameterConverter(private val typeParameterMap: Map<String, String>) :
        PsiTypeVisitor<Any?>() {

        val typeParameterBuilder: TypeParameterBuilder = TypeParameterBuilder.aTypeParameter()

        override fun visitClassType(classType: PsiClassType): Any? {
            val typeConverter = PsiTypeConverter(typeParameterMap)
            val typeBuilder = classType.accept(typeConverter)
            typeParameterBuilder.withType(typeBuilder)

            return super.visitClassType(classType)
        }

        override fun visitWildcardType(wildcardType: PsiWildcardType): Any? {
            wildcardType.bound?.accept(this)

            typeParameterBuilder.withSubTypes(wildcardType.isExtends)
            typeParameterBuilder.withSuperTypes(wildcardType.isSuper)

            return super.visitWildcardType(wildcardType)
        }
    }
}
